beacon-chain/cache/attestation.go

```diff
@@ -0,0 +1,264 @@
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v5/crypto/bls"
+	ethpb "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1/attestation"
+)
+
+type attGroup struct {
+	slot     primitives.Slot
+	local    ethpb.Att
+	external []ethpb.Att
+}
+
+type AttestationCache struct {
+	atts map[attestation.Id]*attGroup
+	sync.RWMutex
+	forkchoiceAtts *forkchoice.Attestations
+}
+
+func NewAttestationCache() *AttestationCache {
+	return &AttestationCache{
+		atts:           make(map[attestation.Id]*attGroup),
+		forkchoiceAtts: forkchoice.New(),
+	}
+}
+
+func (c *AttestationCache) Add(att ethpb.Att) error {
+	if att.IsNil() {
+		return nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	id, err := attestation.NewId(att, attestation.Data)
+	if err != nil {
+		return errors.Wrapf(err, "could not create attestation ID")
+	}
+
+	group := c.atts[id]
+	if group == nil {
+		group = &attGroup{
+			slot: att.GetData().Slot,
+		}
+		c.atts[id] = group
+	}
+
+	if att.IsAggregated() {
+		group.external = append(group.external, att.Clone())
+		return nil
+	}
+
+	local := c.atts[id].local
+	if local == nil {
+		local = att.Clone()
+	}
+	bit := att.GetAggregationBits().BitIndices()[0]
+	if local.GetAggregationBits().BitAt(uint64(bit)) {
+		return nil
+	}
+	sig, err := aggregateSig(local, att)
+	if err != nil {
+		return errors.Wrapf(err, "could not aggregate signatures")
+	}
+
+	local.GetAggregationBits().SetBitAt(uint64(bit), true)
+	local.SetSignature(sig)
+
+	return nil
+}
+
+func (c *AttestationCache) GetAll() []ethpb.Att {
+	c.RLock()
+	defer c.RUnlock()
+
+	var result []ethpb.Att
+	for _, group := range c.atts {
+		if group.local != nil {
+			result = append(result, group.local)
+		}
+		for _, a := range group.external {
+			result = append(result, a)
+		}
+	}
+	return result
+}
+
+func (c *AttestationCache) Count() int {
+	c.RLock()
+	defer c.RUnlock()
+
+	count := 0
+	for _, group := range c.atts {
+		if group.local != nil {
+			count++
+		}
+		count += len(group.external)
+	}
+	return count
+}
+
+func (c *AttestationCache) DeleteCovered(att ethpb.Att) error {
+	if att.IsNil() {
+		return nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	id, err := attestation.NewId(att, attestation.Data)
+	if err != nil {
+		return errors.Wrapf(err, "could not create attestation ID")
+	}
+
+	group := c.atts[id]
+	if group == nil {
+		return nil
+	}
+
+	local := group.local
+	if local != nil {
+		if covered, err := att.GetAggregationBits().Contains(local.GetAggregationBits()); err != nil {
+			return err
+		} else if covered {
+			group.local = nil
+		}
+	}
+
+	attsToKeep := make([]ethpb.Att, 0, len(group.external))
+	for _, a := range group.external {
+		if covered, err := att.GetAggregationBits().Contains(a.GetAggregationBits()); err != nil {
+			return err
+		} else if !covered {
+			attsToKeep = append(attsToKeep, a)
+		}
+	}
+	group.external = attsToKeep
```

Comment:
This is a better approach as it doesn't allocate a new slice 

```suggestion
        idx := 0
	for _, a := range group.external {
		if covered, err := att.GetAggregationBits().Contains(a.GetAggregationBits()); err != nil {
			return err
		} else if !covered {
			group.external[idx] = a
			idx++
		}
	}
	group.external = group.external[:idx]
```